File: siv.ts
// Copyright (C) 2017 Dmitry Chestnykh
// MIT License. See LICENSE file for details.

import { BlockCipher, BlockCipherContructor } from "@stablelib/blockcipher";
import { CMAC, dbl } from "@stablelib/cmac";
import { CTR } from "@stablelib/ctr";
import { wipe } from "@stablelib/wipe";
import { equal } from "@stablelib/constant-time";

/** Maximum number of associated data items */
export const MAX_ASSOCIATED_DATA = 126;

export class SIV {
    _mac: CMAC;
    _ctr: CTR | undefined;
    _macCipher: BlockCipher;
    _encCipher: BlockCipher;
    _tmp1: Uint8Array;
    _tmp2: Uint8Array;

    readonly tagLength: number;

    constructor(cipher: BlockCipherContructor, key: Uint8Array) {
        const macKey = key.subarray(0, key.length / 2 | 0);
        const encKey = key.subarray(key.length / 2 | 0);

        this._macCipher = new cipher(macKey);
        this._encCipher = new cipher(encKey);
        this._mac = new CMAC(this._macCipher);

        if (this._mac.digestLength !== this._mac.blockSize) {
            throw new Error("SIV: this implementation needs CMAC block size to equal tag length");
        }
        this.tagLength = this._mac.digestLength;

        this._tmp1 = new Uint8Array(this._mac.digestLength);
        this._tmp2 = new Uint8Array(this._mac.digestLength);
    }

    seal(associatedData: Uint8Array[], plaintext: Uint8Array, dst?: Uint8Array): Uint8Array {
        if (associatedData.length > MAX_ASSOCIATED_DATA) {
            throw new Error("SIV: too many associated data items");
        }

        // Allocate space for sealed ciphertext.
        const resultLength = this.tagLength + plaintext.length;
        let result;
        if (dst) {
            if (dst.length !== resultLength) {
                throw new Error("SIV: incorrect destination length");
            }
            result = dst;
        } else {
            result = new Uint8Array(resultLength);
        }

        // Authenticate.
        const iv = this._s2v(associatedData, plaintext);
        result.set(iv);

        // Encrypt.
        zeroIVBits(iv);
        this._streamXOR(iv, plaintext, result.subarray(iv.length));
        return result;
    }

    open(associatedData: Uint8Array[], sealed: Uint8Array, dst?: Uint8Array): Uint8Array | null {
        if (associatedData.length > MAX_ASSOCIATED_DATA) {
            throw new Error("SIV: too many associated data items");
        }
        if (sealed.length < this.tagLength) {
            return null;
        }

        // Allocate space for decrypted plaintext.
        const resultLength = sealed.length - this.tagLength;
        let result;
        if (dst) {
            if (dst.length !== resultLength) {
                throw new Error("SIV: incorrect destination length");
            }
            result = dst;
        } else {
            result = new Uint8Array(resultLength);
        }

        // Decrypt.
        const tag = sealed.subarray(0, this.tagLength);
        const iv = this._tmp1;
        iv.set(tag);
        zeroIVBits(iv);
        this._streamXOR(iv, sealed.subarray(this.tagLength), result);

        // Authenticate.
        const expectedTag = this._s2v(associatedData, result);
        if (!equal(expectedTag, tag)) {
            wipe(result);
            return null;
        }
        return result;
    }

    _streamXOR(iv: Uint8Array, src: Uint8Array, dst: Uint8Array) {
        if (!this._ctr) {
            this._ctr = new CTR(this._encCipher, iv);
        } else {
            this._ctr.setCipher(this._encCipher, iv);
        }
        this._ctr.streamXOR(src, dst);
        return dst;
    }

    _s2v(s: Uint8Array[], sn: Uint8Array): Uint8Array {
        if (!s) {
            s = [];
        }

        this._mac.reset();
        wipe(this._tmp1);

        // Note: the standalone S2V returns CMAC(1) if the number of passed
        // vectors is zero, however in SIV contruction this case is never
        // triggered, since we always pass plaintext as the last vector (even
        // if it's zero-length), so we omit this case.
        this._mac.update(this._tmp1);
        this._mac.finish(this._tmp2);
        this._mac.reset();

        for (let i = 0; i < s.length; i++) {
            this._mac.update(s[i]);
            this._mac.finish(this._tmp1);
            this._mac.reset();
            dbl(this._tmp2, this._tmp2);
            xor(this._tmp2, this._tmp1);
        }

        wipe(this._tmp1);

        if (sn.length >= this._mac.blockSize) {
            const n = sn.length - this._mac.blockSize;
            this._tmp1.set(sn.subarray(n));
            this._mac.update(sn.subarray(0, n));
        } else {
            this._tmp1.set(sn);
            this._tmp1[sn.length] = 0x80;
            dbl(this._tmp2, this._tmp2);
        }
        xor(this._tmp1, this._tmp2);
        this._mac.update(this._tmp1);
        this._mac.finish(this._tmp1);
        return this._tmp1;
    }

    clean() {
        wipe(this._tmp1);
        wipe(this._tmp2);
        if (this._ctr) {
            this._ctr.clean();
        }
        this._mac.clean();
        this._encCipher.clean();
        this._macCipher.clean();
        this.tagLength = 0;
    }
}

function xor(a: Uint8Array, b: Uint8Array) {
    for (let i = 0; i < b.length; i++) {
        a[i] ^= b[i];
    }
}

function zeroIVBits(iv: Uint8Array) {
    // "We zero-out the top bit in each of the last two 32-bit words
    // of the IV before assigning it to Ctr"
    //  — http://web.cs.ucdavis.edu/~rogaway/papers/siv.pdf
    iv[iv.length - 8] &= 0x7f;
    iv[iv.length - 4] &= 0x7f;
}

File: ruby/lib/miscreant/siv.rb
# encoding: binary
# frozen_string_literal: true

module Miscreant
  # The SIV misuse resistant authenticated encryption mode
  class SIV
    # Generate a new random AES-SIV key of the given size
    #
    # @param size [Integer] size of key in bytes (32 or 64)
    #
    # @return [String] newly generated AES-SIV key
    def self.generate_key(size = 32)
      raise ArgumentError, "key size must be 32 or 64 bytes" unless [32, 64].include?(size)
      SecureRandom.random_bytes(size)
    end

    # Create a new AES-SIV instance
    #
    # @param key [String] 32-byte or 64-byte Encoding::BINARY cryptographic key
    # @param mac [:CMAC, :PMAC] (optional) MAC function to use (default CMAC)
    def initialize(key, mac_class = :CMAC)
      Internals::Util.validate_bytestring(key, length: [32, 64])
      length = key.length / 2

      @mac = Internals::MAC.const_get(mac_class).new(key[0, length])
      @ctr = Internals::AES::CTR.new(key[length..-1])
    end

    # Inspect this AES-SIV instance
    #
    # @return [String] description of this instance
    def inspect
      to_s
    end

    # Encrypt a message using AES-SIV, authenticating it along with the associated data
    #
    # @param plaintext [String] an Encoding::BINARY string to encrypt
    # @param associated_data [Array<String>] optional array of message headers to authenticate
    #
    # @return [String] encrypted ciphertext
    def seal(plaintext, associated_data = [])
      raise TypeError, "expected String, got #{plaintext.class}" unless plaintext.is_a?(String)
      v = _s2v(associated_data, plaintext)
      ciphertext = @ctr.encrypt(_zero_iv_bits(v), plaintext)
      v + ciphertext
    end

    # Verify and decrypt an AES-SIV ciphertext, authenticating it along with the associated data
    #
    # @param ciphertext [String] an Encoding::BINARY string to decrypt
    # @param associated_data [Array<String>] optional array of message headers to authenticate
    #
    # @raise [Miscreant::IntegrityError] ciphertext and/or associated data are corrupt or tampered with
    # @return [String] decrypted plaintext
    def open(ciphertext, associated_data = [])
      raise TypeError, "expected String, got #{ciphertext.class}" unless ciphertext.is_a?(String)
      v = ciphertext[0, Internals::Block::SIZE]
      plaintext = @ctr.encrypt(_zero_iv_bits(v), ciphertext[Internals::Block::SIZE..-1])
      t = _s2v(associated_data, plaintext)
      raise IntegrityError, "ciphertext verification failure!" unless Internals::Util.ct_equal(t, v)

      plaintext
    end

    private

    # The S2V operation consists of the doubling and XORing of the outputs
    # of the pseudo-random function CMAC.
    #
    # See Section 2.4 of RFC 5297 for more information
    def _s2v(associated_data, plaintext)
      # Note: the standalone S2V returns CMAC(1) if the number of passed
      # vectors is zero, however in SIV construction this case is never
      # triggered, since we always pass plaintext as the last vector (even
      # if it's zero-length), so we omit this case.
      d = Internals::Block.new
      d.xor_in_place(@mac.digest(d.data))

      associated_data.each do |ad|
        d.dbl
        d.xor_in_place(@mac.digest(ad))
      end

      if plaintext.bytesize >= Internals::Block::SIZE
        # TODO: implement this more efficiently by adding IUF support to CMAC
        difference = plaintext.length - Internals::Block::SIZE
        beginning = plaintext[0, difference]
        d.xor_in_place(plaintext[difference..-1])
        msg = beginning + d.data
      else
        d.dbl
        d.xor_in_place(Internals::Util.pad(plaintext, Internals::Block::SIZE))
        msg = d.data
      end

      @mac.digest(msg)
    end

    # "We zero-out the top bit in each of the last two 32-bit words
    # of the IV before assigning it to Ctr"
    # -- http://web.cs.ucdavis.edu/~rogaway/papers/siv.pdf
    def _zero_iv_bits(iv)
      iv = iv.dup
      iv.setbyte(8, iv.getbyte(8) & 0x7f)
      iv.setbyte(12, iv.getbyte(12) & 0x7f)
      iv
    end
  end
end
